Remove dependency from fabric8io/fabric8

rhuss opened the pull request:
I.e. from the helper methods in kubernetes-api. One single dependency to `fabric8io/gitrepo-api` remains which is used by `fabric8:import`. fabric8:import has the most dependencies into the fabric8 CD and app, so I wonder whether it would makes sense to move this into a separate plugin as it is *very* specific to the fabric8 setup. It might also make sense to move out gitrepo-api to an upper level project so that the last dependency would be cut.

There is a *lot* room for optimization, actually quite some code had to be copied over (but astonishingly not so much).

rhuss commented:
@jstrachan @iocanel I wonder how we should proceed with this PR. I guess no one is really keen on a review of 126 changed files :) I tried hard, but of course cannot be sure that something has been broken somewhere. 

There are no significant changes in the logic, I mostly copied things over, but only those which are needed by fmp. Therefore there are also quite many sonarqube issues. The next step would be some code cleanup.

Integration tests are running and I will try to walk through some use cases for every goal tomorrow. 

Since this is still a risk that things are broken, I would like to have a "go" from you :)

rhuss commented:
* `fabric8:setup`, `fabric8:cluster-start`, `fabric8:cluster-stop` works for me after the refactoring
* `fabric8:import` is broken with #950 but this already fails in 3.3.5 and 3.4.0, too (in the same way)

jstrachan commented:
Testing of such huge changes to fmp is non trivial (we kinda need to re-create quickstarts after this has been released using the new release then try it inside fabric8 inside CD pipelines to be sure if anything's broken).

We're heads down trying to get the upstream fabric8 aligned with the latest minishift + openshift.io work for the next week or two - but as part of that work we'll have an end to end test to verify a distribution. So either release these changes a new major beta release (while stopping auto-upgrade of all fabric8 projects to the new major version until we're ready to try it out); or maybe park this PR a couple of weeks?

On `mvn fabric8:import` if the current project is in git then it should only really create a `BuildConfig` which is 100% openshift stuff really and should not depend on anything in openshift thats fabric8 specific; where the fabric8 specific parts come in is detecting gogs/gitlab if there's no git repo and automating the git import. But happy to move it out of fmp if you prefer - given your quest to remove dependencies it sounds like thats inevitable ;)

rhuss commented:
@jstrachan Thanks for the feedback. 

I currently continue to test the refactoring and might be happy to park it e.g. in a next release branch. However, this works really only without very big pain if things on `master` don't change much (and e.g. continue to use dependencies from fabric8io/fabric8 like `Strings.isNotBlank` etc.).

I can rebase such a parked branch for some time if rebasing can be done with a reasonable amount of effort. 

In the meantime, I'm going to continue the manual testing and will branch off my enricher refactoring from this branch if you don't mind.

So, I suggest starting a `4.0` branch for this more aggressive refactoring, cleanups and enhancements. Having those two branches will work only smoothly if we keep some disciplines in our PRs so that rebasing works without too much manual work.

Especially new commits on master must not (or only when absolutely necessary) use

* io.fabric8/kubernetes-api (KubernetesHelper should be ok as I moved it partially over here)
* io.fabric8/fabric8-utils (i replaced with commons-lang3 and guava most of the stuff)
* io.fabric8/fabric8-project-utils

jstrachan commented:
@rhuss totally understood of the fear of parking massive refactors! I'm hoping to not touch fmp at all for the next week or two unless something critical crops up & will try test your refactor ASAP

rhuss commented:
@maxandersen I guess I can close this PR now, as its very likely far too much work to resolve the conflicts.

Actually, is it still a goal to make fabric8-maven-plugin more independent and removing legacy dependencies ?

maxandersen commented:
@pradeepto @hrishin ^^

i think we need to get the work on roadmap and figure out where fmp goes - but in general the less dependent while still provide a way to be compatible with multiple versions of f8/kube  the better ;)

rhuss commented:
I'm reopening this PR and I'm about to rebase it. Should be done soonish.

rhuss commented:
Since we got rid of `fabric8:import` we can remove `fabric8-project` submodule, which makes it even cleaner.

rhuss commented:
@rohanKanojia @nicolaferraro @hrishin @piyush1594 Can we try to get this PR merged as quick is possible before its get rotten again ? I think its really important to get rid of the fabric8io/fabric8 dependency to clean up our legacy.

'would be awesome if you could give it a quick review. I know its not possible to get over all details, and there might be for sure some minor stuff broken (all tests are running so far), but I think its still worth to get this in.

rhuss commented:
@nicolaferraro, unfortunately, I can't really say what has been changed since this PR has started (March 2017). I will cherry pick the change you mentioned and will go through the changes in fabric8 and see whether it might have affected f-m-p. 

However, there is still a risk that we miss a fix or two. But I think its still worth to make this step.

rhuss commented:
I checked all commit in fabric8 since May 2017 and it seems that only your change for the PVs has been added which affects f-m-p.

I readded it.

@hrishin btw, the rt module still depends on fabric8 via kubernetes-assertions, but as we plan to move this out of f-m-p anyway, I left it untouched.

lordofthejars commented on enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/GrafanaLinkEnricher.java:
```diff
@@ -61,28 +61,28 @@ private String findGrafanaLink() {
             query += "&var-project=" + projectName;
         }
-        if (Strings.isNotBlank(version)) {
+        if (StringUtils.isNotBlank(version)) {
```
This is not from this PR, but usually building a URL using string concatenation is a bad idea, and it is always a better approach to use a HttpBuilder where you can generate everything using a DSL approach. So the best approach is to check if any of the libraries used have one or implement one so it can be reused everywhere.

lordofthejars commented on enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/CdEnricher.java:
```diff
@@ -74,7 +72,16 @@ protected boolean getDefaultOnline() {
+        return buildId != null;
+    }
+
+    private String getBuildId() {
```
Is there any static class in charge of doing this inside FMP? I think this logic should be always in a static class so all properties behave with the same approach, and if someday it is decided to change, it is changed everywhere not just in the places we can remember to change.

lordofthejars commented on core/src/test/java/io/fabric8/maven/core/util/KubernetesResourceUtilTest.java:
```diff
@@ -92,7 +91,7 @@ public void containsKind() throws Exception {
         HasMetadata ret = getResource(DEFAULT_RESOURCE_VERSIONING, new File(fabric8Dir, "job.yml"), "app");
         assertEquals("Job", ret.getKind());
-        assertEquals(JOB_VERSION, ret.getApiVersion());
+        assertEquals(KubernetesResourceUtil.JOB_VERSION, ret.getApiVersion());
```
Why use junit asertions when the project already imports assertj which makes writing and reading assertions better. I am not saying to change all assertions, butmaybe the new tests could start using it.

rhuss commented on core/src/main/java/io/fabric8/maven/core/service/ApplyService.java:
```diff
@@ -0,0 +1,1306 @@
+            // for when using vanilla Kubernetes
+            String jenkinshiftUrl = System.getenv("JENKINSHIFT_URL");
+            if (jenkinshiftUrl == null) {
+                jenkinshiftUrl = "http://jenkinshift/";
```
I guess we get rid of JenkinShift soon anyway (can't even remember what is was used for), but I completely agreed with you.
To set the expectation right for this PR, its not about cleaning up the code, its about removing the ugly dependency first. So some parts of fabric8 had to be copied over literally.

Cleaning up will be one of the next steps.
